VMModel error handling improved and tests added

vhavrylov-ncube opened the pull request.

vhavrylov-ncube committed:
VMModel runtime errors improved, tests added

vhavrylov-ncube committed:
Model error handling improved
- Meaningful VM->RuntimeErrors are raised wherever possible
- std::exceptions are thrown if an error can not be handled in-place.

vhavrylov-ncube committed:
Added unit test for an empty Sequential model

baykaner commented on libs/vm-modules/include/vm_modules/ml/model/model.hpp:
```diff
@@ -118,6 +118,8 @@ class VMModel : public fetch::vm::Object
   ModelCategory      model_category_ = ModelCategory::NONE;
   bool               compiled_       = false;
 
+  static constexpr size_t min_allowed_hidden_layers_ = 2;
```
this value should be 1; and you should fix the code that doesn't correctly handle that case

baykaner replied:
also use std::size_t (code style)

vhavrylov-ncube replied:
I would keep it 2 for now as it protects user from a runtime crash. Indeed we need to fix Graph code, but right now I can not estimate if that fix is small or huge (and what risk does it add), so I guess it might be better to create a separate task for this investigation, and set VMModel::min_allowed_hidden_layers_ to 1 in that scope.

vhavrylov-ncube replied:
I leave a comment above this constant to explain the current value

vhavrylov-ncube replied:
After clarification with @MissingNO57 : no changes are required to existing code, it works fine, just a naming confusion. I have renamed necessary constants and variables.

baykaner commented on libs/vm-modules/include/vm_modules/ml/model/model.hpp:
```diff
@@ -101,7 +101,7 @@ class VMModel : public fetch::vm::Object
   static void Bind(fetch::vm::Module &module);
 
-  ModelPtrType &GetModel();
+  void SetUnderlyingModelInstance(ModelPtrType const &instance);
```
just SetModel is fine:

also why do you remove the getter?

vhavrylov-ncube replied:
The only usage of this getter was in this file, to make assignments to its return value. This approach

a.getSomething() = b;

discards constness of the getter, and I guess it would be better to have explicitly non-constant setter.

baykaner commented on libs/vm-modules/src/ml/model/model.cpp:
```diff
@@ -161,19 +157,17 @@ void VMModel::CompileSequential(fetch::vm::Ptr<fetch::vm::String> const &loss,
 {
-  OptimiserType const optimiser_type = ParseName(optimiser->string(), optimisers_, "optimiser");
-  if (optimiser_type != OptimiserType::ADAM)
+  size_t const total_hidden_layers = in_layers->elements.size();
```
std::size_t

baykaner replied:
also calling elements.size() is very cheap - you don't need to store it to a variable. but it's fine if you think it's clearer to read that way

vhavrylov-ncube replied:
Yeap, it's more about naming than cacheing :)

baykaner commented on libs/vm-modules/tests/unit/ml_model_tests.cpp:
```diff
@@ -347,6 +347,140 @@ TEST_F(VMModelTests, model_init_with_wrong_name)
   EXPECT_THROW(toolkit.Run(), std::runtime_error);
 }
 
+TEST_F(VMModelTests, model_add_invalid_layers)
```
this test has nice coverage, but it would be preferable to wrap up the re-usable code in the test fixture, and then have multiple small tests; that way when one fails you immediately know which invalid case is at fault

vhavrylov-ncube replied:
Agreed

vhavrylov-ncube committed:
LayerCount() getter added for ML Model
+ code style fixed
+ naming fixes
+ added constness to some Model<> getters

vhavrylov-ncube committed:
Reverted changes to ML-MODEL<>

+ only Sequential Model is touched with layer counting

vhavrylov-ncube committed:
Hidden layers -> layer shapes reanming
To prevent confusion

vhavrylov-ncube committed:
PR fixes

- huge unit test splitted to separate ones
- naming corrected for Model creation

baykaner commented on libs/ml/include/ml/model/sequential.hpp:
```diff
@@ -45,12 +45,14 @@ class Sequential : public Model<TensorType>
 private:
-  SizeType    layer_count_ = 0;
   std::string prev_layer_;
+  SizeType    layer_count_ = 0;
```
if you add any new member variables to an object, this will need to be reflected in the object serialisation and deserialistion. Please speak to me, Jiri, or Emma if you're not sure exactly how to do that.

vhavrylov-ncube replied:
Ok, but this variable actually already exists in
https://github.com/fetchai/ledger/blob/master/libs/ml/include/ml/model/sequential.hpp
and there are several lines of code using it.
I don't know why Github highlighted this line as a change, becase I've added a getter method, not a member.

vhavrylov-ncube replied:
Serialisation/deserialisation of this member is already implemented in sequential.hpp, so no action required.

baykaner commented on libs/vm-modules/src/ml/model/model.cpp:
```diff
@@ -137,72 +135,96 @@ Ptr<VMModel> VMModel::Constructor(VM *vm, TypeId type_id,
-  for (size_t i = 0; i < n_elements; ++i)
+  std::vector<math::SizeType> shapes;
+  shapes.reserve(total_layer_shapes);
+  for (size_t i = 0; i < total_layer_shapes; ++i)
```
std::size_t (style)

vhavrylov-ncube replied:
oh sorry for that :)

baykaner commented on libs/vm-modules/src/ml/model/model.cpp:
```diff
@@ -395,29 +417,34 @@ void VMModel::AssertLayerTypeMatches(SupportedLayerType                layer,
-                              "\" layer."};
-    vm_->RuntimeError(message);
-    throw std::runtime_error(message);
+    throw std::runtime_error("Invalid params specified for \"" + LAYER_NAMES_.at(layer) +
```
I think this was better before. The reason is that if we try to run a smart contract that uses this model function and fails here, the vm->RuntimeError(message) should route that useful message to std::out for the developer in a sensible way - whereas merely throwing will not give this information to the developer.

baykaner replied:
if you like we can discuss this next week and I can show you how to run a simple smart contract and try it out for yourself. That might make this clearer.

vhavrylov-ncube replied:
Runtime error will be called in a caller function, the top level .AddLayer<> contains a try-catch block that now covers all possible usages of ParseName, Assert.. and AddLayerSpecificImpl, so any exception will be caught and turned into VM-runtime error. So we still have VM logs with this error if occurs.

vhavrylov-ncube replied:
I have replaced RuntimeErrors with exceptions wherever an immediate stop of the machine does not happen, e.g. where the control flow does not return to VM handlers. And vice-versa, in AddLayer and other etch-object-member-functions I replaced c++ exceptions with RuntimeErrors - because after return from this method control flow goes to VM hadler, the machine can be safely stopped and error log appears. Mixing both things in one code block did not provide expected result for me.

baykaner commented on libs/vm-modules/src/ml/model/model.cpp:
```diff
@@ -395,29 +417,34 @@ void VMModel::AssertLayerTypeMatches(SupportedLayerType                layer,
-    std::string const message{"No \"add\" method exists for non-sequential models!"};
-    vm_->RuntimeError(message);
-    throw std::runtime_error(message);
+    throw std::runtime_error("Layer adding is allowed only for sequential models!");
```
same as above

vhavrylov-ncube replied:
This function is also always covered by catch blocks, either AddLayer<> or Compile... one, so the error message will be shown neatly.

baykaner commented on libs/vm-modules/tests/unit/ml_model_tests.cpp:
```diff
@@ -32,11 +32,52 @@ namespace {
+
+  void TestInvalidLayerCompilation(std::vector<std::string> sources)
+  {
+    for (const auto &test_case : sources)
```
still you could ditch this for-loop, take in just a single source, and separate all the tests for the most clarity

vhavrylov-ncube replied:
Ok, I'll do it

baykaner commented on libs/vm-modules/tests/unit/ml_model_tests.cpp:
```diff
@@ -32,11 +32,52 @@ namespace {
+  {
+    for (const auto &test_case : sources)
+    {
+      std::cout << "Testing invalid layer: " << test_case << std::endl;
```
then you wouldn't need to print this too

vhavrylov-ncube replied:
done

baykaner reviewed (changes requested):
tests look much cleaner, but still I would separate them as much as possible. Also you must implement serialisation/deserialisation for the layer count.

vhavrylov-ncube commented on libs/vm-modules/src/ml/model/model.cpp:
```diff
@@ -395,29 +417,34 @@ void VMModel::AssertLayerTypeMatches(SupportedLayerType                layer,
+  {
+    SupportedLayerType const layer_type = ParseName(layer->string(), layer_types_, "layer type");
+    AddLayerSpecificImpl(layer_type, args...);
+    compiled_ = false;
```
This try block covers all Assertions and ParseName's, that could be invoked in AddLayerSpecificImpl(), so if assignment [compiled_ = false] takes place, it is guaranteed that a layer was indeed added.

vhavrylov-ncube committed:
Unit tests splitted, lint corrections

fetch-bot committed:
Merge branch 'master' into refactor/ml-219-graph-etch-bindings
